helix-vcs/src/diff/worker.rs

```diff
@@ -0,0 +1,196 @@
+        self.new_hunks.clear();
+    }
+
+    fn perform_diff(&mut self, input: &InternedInput<RopeSlice>) {
+        imara_diff::diff(ALGORITHM, input, |before: Range<u32>, after: Range<u32>| {
+            self.new_hunks.push(Hunk { before, after })
+        })
+    }
+}
+
+struct EventAccumulator<'a> {
+    diff_base: Option<Rope>,
+    doc: Option<Rope>,
+    render_stratagey: RenderStrategy,
+    redraw_handle: &'a RedrawHandle,
+    render_lock: Option<RwLockReadGuard<'a, ()>>,
+    timeout: Instant,
+}
+
+impl<'a> EventAccumulator<'a> {
+    fn new(redraw_handle: &'a RedrawHandle) -> EventAccumulator<'a> {
+        EventAccumulator {
+            diff_base: None,
+            doc: None,
+            render_stratagey: RenderStrategy::Async,
+            render_lock: None,
+            redraw_handle,
+            timeout: Instant::now(),
+        }
+    }
+
+    async fn handle_event(&mut self, event: Event) {
+        let dst = if event.is_base {
+            &mut self.diff_base
+        } else {
+            &mut self.doc
+        };
+
+        *dst = Some(event.text);
+
+        // always prefer the most synchronus requested render mode
+        if event.render_strategy > self.render_stratagey {
+            if self.render_lock.is_none() {
+                self.timeout = Instant::now() + Duration::from_millis(SYNC_DIFF_TIMEOUT);
+                self.render_lock = Some(self.redraw_handle.1.read().await);
+            }
+            self.render_stratagey = event.render_strategy
+        }
+    }
+
+    async fn accumulate_debounced_events(
+        &mut self,
+        channel: &mut UnboundedReceiver<Event>,
+        redraw_handle: RedrawHandle,
+        diff_finished_notify: Arc<Notify>,
+    ) {
+        let async_debounce = Duration::from_millis(DIFF_DEBOUNCE_TIME_ASYNC);
+        let sync_debounce = Duration::from_millis(DIFF_DEBOUNCE_TIME_SYNC);
+        loop {
+            let debounce = if self.render_stratagey == RenderStrategy::Async {
+                async_debounce
+            } else {
+                sync_debounce
+            };
+
+            if let Ok(Some(event)) = timeout(debounce, channel.recv()).await {
+                self.handle_event(event).await;
+            } else {
+                break;
+            }
+        }
+
+        // setup task to trigger the rendering
+        // with the choosen render stragey
+        match self.render_stratagey {
+            RenderStrategy::Async => {
+                tokio::spawn(async move {
+                    diff_finished_notify.notified().await;
+                    redraw_handle.0.notify_one();
+                });
+            }
+            RenderStrategy::SyncWithTimeout => {
+                let timeout = self.timeout;
+                tokio::spawn(async move {
+                    let res = {
+                        // Aquire a lock on the redraw handle.
+                        // The lock will block the rendering from occuring while held.
```

Comment:
```suggestion
                        // Acquire a lock on the redraw handle.
                        // The lock will block the rendering from occurring while held.
```

and also in the RenderStrategy::Sync block below

Reply:
Its super late in my timezone right now, I just applied your fixes on my phone. Will fix this in a few hours once I wake up

Reply:
fixed this now